### ISSUE TITLE
Transmuting the same trait object with different lifetimes

### ISSUE THREAD

**Issue by Darksonn** [rust-lang/unsafe-code-guidelines#282]
Consider the following situation:

 * You have an `Box<dyn SomeTrait + 'a>`
 * You transmute it to a `Box<dyn SomeTrait + 'static>`

Is this transmute UB, assuming you actually destroy the box before the end of the lifetime `'a`? It is a transmute that I see people do relatively often.

Similarly for `&dyn SomeTrait` and `&mut dyn SomeTrait`.

**Comment by bjorn3** [comment]
Transmuting `&dyn SomeTrait` to `&mut dyn SomeTrait` is UB as far as I know. Transmuting lifetimes is not.

**Comment by RalfJung** [comment]
I don't think we currently allow any kind of transmutes on `dyn Trait`. So this is not UB but also not guaranteed to give a reasonable result; it exploits unspecified implementation details (akin to transmuting `repr(Rust)` structs).

But this seems like a rather reasonable one, assuming the underlying data is indeed only accessed during the lifetime `'a`.

**Comment by Darksonn** [comment]
@bjorn3 Right. What I meant was `&dyn SomeTrait + 'a` → `&dyn SomeTrait + 'static` and similarly for mutable references.

@RalfJung "So this is not UB"? Would exploiting unspecified implementation details not be UB?

**Comment by RalfJung** [comment]
>  Would exploiting unspecified implementation details not be UB?

No, not if you guessed those details correctly ;) . UB is documented at https://doc.rust-lang.org/reference/behavior-considered-undefined.html.
I might be language lawyering here, but I think this is an important distinction. Exploiting unspecified implementation details can easily lead to UB if you get them wrong or if they change, but if you happen to get them right, your code doesn't have UB. It's still a bug and outside the realms of what Rust guarantees, but it's distinct from UB.

**Comment by RalfJung** [comment]
Mechanically the question boils down to whether the vtable is still valid for the other trait. I don't actually know whether vtables can be affected by that lifetime in the dyn trait type -- @lcnr can you help us out here?

Also @Darksonn I know a bunch of stuff happened since this question in the area of raw pointer dyn trait casts and transmutes; do you have an update for this?

**Comment by lcnr** [comment]
> Mechanically the question boils down to whether the vtable is still valid for the other trait. I don't actually know whether vtables can be affected by that lifetime in the dyn trait type -- @lcnr can you help us out here?

Free regions generally do not impact behavior, which should imply that `Box<dyn SomeTrait + 'a>` and `Box<dyn SomeTrait + 'static>` have the exact same behavior. ofc get UB using safe code after this transmute, but i believe the transmute itself should be fine

I expect that we already guarantee this for ADTs like `Foo<'a>` to `Foo<'b>` 🤔

**Comment by lcnr** [comment]
we compute vtables for trait objects even after erasing the higher ranked regions of the trait object, so `for<'a> dyn Trait<'a>` and `dyn Trait<'static>` also get the same vtable. `dyn Trait<for<'a> fn(&'a ())>` and `dyn Trait<fn(&'static ()>` do have to be different types however

**Comment by Darksonn** [comment]
1. The trait objects use vtables of the same layout, so there is nothing wrong from that perspective.
2. There might be methods that it would be UB to call (e.g. methods with a `where Self: 'static` trait bound).
3. https://github.com/rust-lang/rust/issues/141402 suggests exactly this transmute to end-users.

Note that I'm only talking about trailing lifetimes. Other kinds of lifetimes are a whole different story; as @lcnr points out, those may actually have completely different vtables.

**Comment by RalfJung** [comment]
> I expect that we already guarantee this for ADTs like Foo<'a> to Foo<'b> 

I don't think we guarantee anything like that.

> Other kinds of lifetimes are a whole different story

Yeah I think that is what I was remembering.

However, that issue also says "Casting these parts of trait objects can invalidate the VTable for the trait object, allowing dispatching to methods that should not be callable" about point 1 which is "Casting *const dyn Trait + 'a to *const dyn Trait + 'b". Doesn't that mean the transmute can be wrong? If not, then why do we disallow the cast...?

Cc @BoxyUwu

**Comment by lcnr** [comment]
>  If not, then why do we disallow the cast...?

we disallow the cast as it is safe, but allows calling methods whose where-bounds don't hold, resulting in UB using only safe code.

As a transmute is unsafe, it's the users fault if they then call a method on the trait object whose where-bounds aren't actually satisfied by the underlying type

**Comment by RalfJung** [comment]
Ah right we can do dyn method dispatch with raw pointer receivers, I keep forgetting that.

**Comment by Darksonn** [comment]
A vtable contains a few different kinds of data. Currently:

1. Size and alignment of the type.
2. Destructor of the type.
3. Function pointers for each function defined by the trait.
4. For each super trait, a pointer to the vtable of said super trait.
5. Maybe also the type name and other stuff like that.

And when you transmute from `dyn Trait + 'short` to `dyn Trait + 'long` you have a situation where although the two types have compatible vtables, there may be some entries in the vtable that are illegal to use for an `dyn Trait + 'short`, but legal to use for an `dyn Trait + 'long`. And so, even if performing the transmute itself is not UB, using any of those illegal vtable entries is UB. (And using said illegal vtable entry may be a safe operation!)

The interesting part here is that the above list is growing ... we recently added (4.) when trait upcasting was added to the language. This means that, even if transmuting `dyn Trait + 'a` to `dyn Trait + 'static` results in a vtable where all entries are legal today, we could, in the future, add additional vtable entries so that this transmute results in a vtable containing an illegal entry that is required to be legal for an `dyn Trait + 'static`.

So for these reasons, I would say that these transmutes will, at the very least, violate safety invariants. Whether they trigger language UB depends on how the resulting trait object is used, and "dropped before end of `'a`" is not a sufficient condition to avoid triggering language UB, depending on what's in the vtable. Precisely what we should guarantee here, if anything, is not super clear to me.

> Ah right we can do dyn method dispatch with raw pointer receivers, I keep forgetting that.

You can also do trait upcasting on raw pointers, which honestly is no different from dyn method dispatch here.

**Comment by RalfJung** [comment]
> And when you transmute from dyn Trait + 'short to dyn Trait + 'long you have a situation where although the two types have compatible vtables, there may be some entries in the vtable that are illegal to use for an dyn Trait + 'short, but legal to use for an dyn Trait + 'long. 

We generally require the vtable to be valid for the type, even for raw pointers. So this sounds like the transmute would be UB since it can violate a language/validity invariants.

**Comment by Darksonn** [comment]
Yes I agree, but I could imagine that we want to make it a safety invariant only, so that you can perform the transmute in limited cases as long as you only perform a certain limited set of actions on the transmuted trait object.

For example, maybe you are trying to implement scoped thread spawn in terms of a `'static`-only thread spawn primitive. That probably involves casting `dyn FnOnce() + 'a` to `dyn FnOnce() + 'static`.

**Comment by Darksonn** [comment]
Another example of this being useful is [`BroadcastStream`](https://docs.rs/tokio-stream/latest/tokio_stream/wrappers/struct.BroadcastStream.html) from tokio-stream. This type does not really need the `T: 'static` bound. But I don't think there's any way to remove the `T: 'static` bound from this type without casting `dyn Future + 'a` to `dyn Future + 'static`.

Well, that's not quite true - you can redefine the type to have a lifetime: `BroadcastStream<'a, T>` and relax the bound to `T: 'a`. But there's no way to do it when the type is defined simply as `BroadcastStream<T>` with no useless lifetime parameter.

I'm pretty sure this is what motivated originally asking this question.

**Comment by zachs18** [comment]
> For example, maybe you are trying to implement scoped thread spawn in terms of a `'static`-only thread spawn primitive. That probably involves casting `dyn FnOnce() + 'a` to `dyn FnOnce() + 'static`.

The stdlib does almost exactly this: transmuting `*mut (dyn FnOnce() + Send + '_)` to `*mut (dyn FnOnce() + Send + 'static)` to convert `Box<dyn FnOnce() + Send + '_>` to `Box<dyn FnOnce() + Send + 'static>` [in `spawn_unchecked`](https://std.noratrieb.dev/x86_64-unknown-linux-gnu/src/std/thread/lifecycle.rs.html#110), which is used to implement std's scoped thread API.

`crossbeam` also does this in its [scoped thread API](https://docs.rs/crossbeam-utils/0.8.21/src/crossbeam_utils/thread.rs.html#464), transmuting `Box<dyn FnOnce() + Send + 'env>` to `Box<dyn FnOnce() + Send + 'static>` directly.

**Comment by RalfJung** [comment]
FWIW it seems Miri currently entirely ignores the "region" field -- which I guess also means the compiler doesn't actually need that region field to compute the vtable? Like, [this code](https://github.com/rust-lang/rust/blob/26f3337d4eda0ba22b615744fda0185d0ee344b1/compiler/rustc_const_eval/src/interpret/traits.rs#L67-L103) doesn't need it at all.

**Comment by lcnr** [comment]
> The interesting part here is that the above list is growing ... we recently added (4.) when trait upcasting was added to the language. This means that, even if transmuting `dyn Trait + 'a` to `dyn Trait + 'static` results in a vtable where all entries are legal today, we could, in the future, add additional vtable entries so that this transmute results in a vtable containing an illegal entry that is required to be legal for an `dyn Trait + 'static`.

> FWIW it seems Miri currently entirely ignores the "region" field -- which I guess also means the compiler doesn't actually need that region field to compute the vtable? Like, [this code](https://github.com/rust-lang/rust/blob/26f3337d4eda0ba22b615744fda0185d0ee344b1/compiler/rustc_const_eval/src/interpret/traits.rs#L67-L103) doesn't need it at all.


Lifetime constraints do not impact codegen/method selection. The current implementation definitely cannot distinguish between `dyn Trait + 'a` and `dyn Trait + 'static` and I feel quite confident that we never will.

The question of "given a trait object of type `dyn Trait + 'static`, could the compiler in theory emit implicit behavior which relies on the object being valid for `'static`" is 🤷 I don't see what that would be. We also don't consider lifetimes during MIR building so we should definitely not get any instances of "compiles with both `+ 'static` and `+ 'a` but only sound with `'a`".
